devicecontrol.cpp

```diff
@@ -11,28 +11,90 @@
 */
 
 #include "devicecontrol.h"
-
 #include "device_adaptor.h"
 
+#include "hybrisprovider.h"
+
 #include <QtDBus/QDBusVariant>
+#include <dbus/dbus.h>
 
-DeviceControl::DeviceControl(QObject *parent)
-:   QObject(parent), m_powered(false)
+DeviceControl::DeviceControl(HybrisProvider *parent)
+:   QObject(parent), m_provider(0), m_enabled(false)
 {
     new DeviceAdaptor(this);
 }
 
-bool DeviceControl::powered() const
+bool DeviceControl::enabled() const
 {
-    return m_powered;
+    return m_enabled;
 }
 
-void DeviceControl::setPowered(bool powered)
+void DeviceControl::setEnabled(bool newEnabled)
 {
-    if (m_powered == powered)
+    if (m_enabled == newEnabled)
         return;
 
-    m_powered = powered;
-    emit poweredChanged();
-    emit PropertyChanged(QStringLiteral("Powered"), QDBusVariant(powered));
+    if (QDBusContext::calledFromDBus()) {
+        QDBusConnection conn = connection();
+        DBusConnection *internalConnection = static_cast<DBusConnection*>(conn.internalPointer());
+        unsigned long dbusRemotePid = 0;
+        dbus_bool_t gotPid = dbus_connection_get_unix_process_id(internalConnection, &dbusRemotePid);
+        if (!gotPid) {
+            return; // cannot determine pid of caller.
+        } else if (!applicationIsPermitted(dbusRemotePid)) {
+            return; // only connman is allowed to use this API.
+        }
+        if (m_provider) {
+            // connman has disabled/enabled gps due to flight mode enter/exit.
+            m_provider->setGpsEnabled(newEnabled);
+        }
+    }
+
+    m_enabled = newEnabled;
+    emit enabledChanged();
+    emit PropertyChanged(QStringLiteral("Enabled"), QDBusVariant(newEnabled));
+}
+
+bool DeviceControl::applicationIsPermitted(unsigned long pid) const
+{
+    if (pid == 0) {
+        return true;
+    }
+
+    const QString pidFile = QStringLiteral("/proc/%1/cmdline").arg(pid);
+    if (!QFile::exists(pidFile)) {
+        return false;
+    }
+
+    QFile file(pidFile);
+    if (!file.open(QIODevice::ReadOnly | QIODevice::Text)) {
+        return false;
+    }
+
+    QByteArray contents;
+    char data;
+    while (file.read(&data, 1)) {
+        if (data == '\0') {
+            data = ' ';
+        }
+        contents.append(data);
+    }
+
+    const QString applicationId(QString::fromUtf8(contents).trimmed());
+    return applicationId.contains(QStringLiteral("connmand"));
```

Comment:
This is easy to bypass and as at the moment connman is run as root, it shouldn't be even needed.

Or furthermore, there's com.jollamobile.gps.conf that already should allow only root, so it appears to me all the policy checks here are unnecessary. If connman is needed to be run as some other user, it's always better if that can be done with dbus policy files instead of hard-coding. Elsewhere we do have some some c++ based checks because policy file doesn't support checking privileged group.